src/design-system/components/menu/Menu.tsx

```diff
@@ -0,0 +1,196 @@
+import { FC, useState } from 'react'
+import { Menu as MuiMenu, PopoverOrigin } from '@mui/material'
+
+import {
+  MenuItemProps as CommonMenuItemProps,
+  OnItemClickArgs
+} from '~/design-system/components/menu-item/MenuItem.types'
+import MenuItem from '../menu-item/MenuItem'
+import {
+  MenuItemColorVariant,
+  MenuItemVariant
+} from '../menu-item/MenuItem.constants'
+
+import '~scss-components/menu/Menu.scss'
+
+const removeAllItemsTitle = 'Clear all'
+const defaultNoItemsMessage = 'No items.'
+
+interface NestedMenuItemProps extends CommonMenuItemProps {
+  defaultOnItemClickArgs?: OnItemClickArgs
+  isInitiallyToggled?: boolean
+}
+
+interface MenuItemProps extends NestedMenuItemProps {
+  nestedMenuItems?: NestedMenuItemProps[]
+}
+
+interface MenuProps {
+  anchorEl: HTMLElement | null
+  setAnchorEl: (anchorEl: HTMLElement | null) => void
+  menuItems: MenuItemProps[]
+  allowToggleMultipleItems?: boolean
+  anchorOrigin?: PopoverOrigin
+  density?: 1 | 2
+  defaultOnItemClick?: (args: OnItemClickArgs) => void
+  isItemsRemovalEnabled?: boolean
+  noItemsMessage?: string
+  maxHeight?: number
+  minWidth?: number
+  transformOrigin?: PopoverOrigin
+  toggledItemsTitles?: string[]
+  onToggleItemsChange?: (newTitles: string[]) => void
+}
+
+const Menu: FC<MenuProps> = ({
+  anchorEl,
+  setAnchorEl,
+  menuItems,
+  defaultOnItemClick,
+  maxHeight,
+  minWidth,
+  noItemsMessage = defaultNoItemsMessage,
+  density = 1,
+  allowToggleMultipleItems = false,
+  isItemsRemovalEnabled = false,
+  toggledItemsTitles: customToggledItemsTitles,
+  onToggleItemsChange,
+  ...menuProps
+}: MenuProps) => {
+  const [items, setItems] = useState<MenuItemProps[]>(menuItems)
+  const [internalToggledItemsTitles, setInternalToggledItemsTitles] = useState<
+    string[]
+  >(
+    allowToggleMultipleItems
+      ? menuItems
+          .filter((item) => item.isInitiallyToggled)
+          .map((item) => item.title)
+      : []
+  )
+
+  const toggledItemsTitles =
+    customToggledItemsTitles && onToggleItemsChange
+      ? customToggledItemsTitles
+      : internalToggledItemsTitles
+
+  const setToggledItemsTitles = (
+    updater: (prevItems: string[]) => string[]
+  ) => {
+    onToggleItemsChange
+      ? onToggleItemsChange(updater(toggledItemsTitles))
+      : setInternalToggledItemsTitles(updater)
+  }
+
+  const toggleAsSingleItem = (itemTitle: string) => {
+    setToggledItemsTitles((previousItems) =>
```

Comment:
If the arrow function doesn't fit into a single line - add curly braces. It will be a lot more readable with them.